ethcore/src/snapshot/service.rs

```diff
@@ -695,8 +721,9 @@ impl Service {
 			Ok(()) |
 			Err(Error::Snapshot(SnapshotError::RestorationAborted)) => (),
 			Err(e) => {
+				// TODO: after this we're sometimes deadlocked
```

Comment:
We should actually always deadlock - because we already have `self.restoration.lock()` acquired in line 720. parking_lot locks are not re-entrant.

Instead of calling `abort_restore()` we should rather use `*restoration = None` here. and maybe add a trace line if you care about it.
Alternatively we could have `abort_restore` take the locked `restoration`. IMHO if we do that for  some methods already we should do that for all of them.

Reply:
> We should actually always deadlock

Well, for some reason we do not. It's pretty rare, I've seen it twice.

> Instead of calling abort_restore() we should rather use *restoration = None

That's what `abort_restore()` does though: [here](https://github.com/paritytech/parity-ethereum/blob/72c58b5319d1d7d9299f7aeded0ff4e6043e38fa/ethcore/src/snapshot/service.rs#L853). Why is it better to do it here?

Reply:
> Why is it better to do it here?

To avoid locking twice?

As I said the second ption should be `abort_restore_with_restoration(locked_res: &mut Option<Restoration>)` so we can keep the logic there, but just pass the existing lock.

Reply:
So the reason I removed `*restoration = None` from here and replaced it with a call to `abort_restore()` was to collect all shutdown actions in one place. Not that `abort_restore()` is doing very much atm, but I still think it's good to have a single point of abortion.
is locking twice really a concern though: this is an error handler and I don't think we care much about performance?

> `abort_restore_with_restoration(locked_res…)`

The `restoration` is [a member of the struct](https://github.com/paritytech/parity-ethereum/blob/72c58b5319d1d7d9299f7aeded0ff4e6043e38fa/ethcore/src/snapshot/service.rs#L234) here, so I'm not sure what the point of passing it as a param tbh. Now you got me confused!

Reply:
Since `parking_lot::Mutex` is not re-entrant locking it twice in the same thread can (or does) lead to a deadlock. So the call like that:
```
fn some(&self) {
  let mut restorating = self.restoration.lock();
  ....
  self.abort_restore();
}

fn abort_restore(&self) {
  *self.restoration.lock() = None
}
```
is guaranteed to deadlock, I thought always, but I suspect it might be just random.

In the same file I suppose we already had deadlock issues, so the `with_restoration` pattern emerged, where instead of locking resources locally in a particular function you get passed a (mutable) reference to the locked resource:
```
let mut restoration = self.restoration.lock();
self.feed_chunk_with_restoration(&mut restoration)
```
I propose to use exactly the same patter for `abort_restore` - I'm totally fine grouping all the shutdown actions there, but to prevent potential double-locking and potential deadlocks we can pass the locked resource, so the first example becomes:

```
fn some(&self) {
  let mut restoration = self.restoration.lock();
  ....
  self.abort_restore(&mut restoration);
}

fn abort_restore(&self, res: &mut Option<Restoration>) {
  *res = None
}

Reply:
I see, thank you for explaining.

Is this an option?

```
		let r = {
			let mut restoration = self.restoration.lock();
			self.feed_chunk_with_restoration(&mut restoration, hash, chunk, is_state)
		};
		match r {
			…
			Err(e) => {
				…
				self.abort_restore();
				…
			}
		}

```

Reply:
According to [docs](https://amanieu.github.io/parking_lot/parking_lot/struct.Mutex.html#method.lock):
> Attempts to lock a mutex in the thread which already holds the lock will result in a deadlock.

So I'm not sure either why it didn't always deadlock. The proposed fix looks good.